java-checks/src/main/java/org/sonar/java/checks/StringConcatToTextBlockCheck.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S6126")
+public class StringConcatToTextBlockCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  private static final String MESSAGE = "Replace this String concatenation with Text block.";
+  private static final Set<Tree> VISITED_NODES = new HashSet<>();
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava15Compatible();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.PLUS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    VISITED_NODES.add(tree);
+    if (VISITED_NODES.contains(tree.parent())) {
+      return;
+    }
+    Object constant = ExpressionUtils.resolveAsConstant(((ExpressionTree) tree));
+    if (constant instanceof String) {
+      String value = (String) constant;
+      if (value.split("\\\\n").length > 2) {
+        reportIssue(tree, MESSAGE);
+      }
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    VISITED_NODES.remove(tree);
```

Comment:
If there's an exception during the analysis of one file, the analysis continues on the next file, and methods like `leaveNode(Tree tree)` or `leaveFile(JavaFileScannerContext context)` are not called.
That why we use this pattern that ensure that everything is clear at the beginning of the analysis of one file:
```
  @Override
  public void setContext(JavaFileScannerContext context) {
    visitedNodes.clear();
    super.setContext(context);
  }
```
Then, I don't think it's useful to remove the tree each time we leave the node, the global clean will be enough.